csrc/optimization/optimize_consecutive_cast.cpp

```diff
@@ -0,0 +1,108 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <ir_utils.h>
+#include <optimization/opt_pass.h>
+
+namespace nvfuser::optimization {
+
+namespace {
+
+class ConsecutiveCastPass : OptimizationPass {
+ public:
+  static void runPass(Fusion* fusion) {
+    auto is_cast_op = [](Expr* expr) {
+      if (expr->isA<UnaryOp>()) {
+        auto op = expr->as<UnaryOp>();
+        if (op->getUnaryOpType() == UnaryOpType::Cast) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    // NOTE: not the most efficient pass
+    std::unordered_map<Val*, Val*> replacement_map;
+    for (auto expr : fusion->exprs()) {
+      if (is_cast_op(expr)) {
+        bool mutated = false;
+        while (true) {
+          // in the loop, we just repetitively skip consecutive casts.
+          auto intermediate_cast = expr->input(0);
+          auto prev_expr = intermediate_cast->definition();
+          if (prev_expr != nullptr && is_cast_op(prev_expr)) {
+            auto original_dtype = prev_expr->input(0)->getDataType().value();
+            auto intermediate_dtype = intermediate_cast->getDataType().value();
+            auto out_dtype = expr->output(0)->getDataType().value();
+            // cases where skipping the intermediate cast is relatively safe,
+            // two conditions:
+            //   1. original_dtype is the same as out_dtype; or
+            //   2. we support direct cast from original_dtype to out_dtype.
+            // and
+            //   1. intermediate_dtype is the same type category as with
```

Comment:
Can this be omitted? For example, I want to debug AMP and emulate fp16 to learn about the numerics of it to figure out why I am getting an inf. Probably I would want to do cast fp32->fp16->fp32 and run followup ops in fp32 to see whether mathematically the large value is due to rounding error or just due to the operation itself is just unstable. I think if we optimize this out, the user will get an no-op.